test.l

```diff
@@ -1,31 +1,34 @@
 #!/usr/bin/env bin/lumen
 
+(when-compiling
+  (define-macro test (x msg)
+    `(if (not ,x)
+         (do (set failed (+ failed 1))
+             (return ,msg))
+       (inc passed)))
+
+  (define-macro test= (a b)
+    (let-unique (x y)
+      `(let (,x ,a ,y ,b)
+         (test (equal? ,x ,y)
+               (cat "failed: expected " (str ,x) ", was " (str ,y))))))
+
+  (define-macro define-test (name rest: body)
+    `(add tests (list ',name (fn () ,@body))))
+
+  nil)
```

Comment:
`when-compiling` allows you to compute expressions at compile time. As a trivial example, 

```
(when-compiling
  `(print ,(cat "h" "i")))
```

is equivalent to

```
(print "hi")
```

The `nil` prevents `when-compiling` from generating any output, since

`(when-compiling nil)`

is the same as writing nothing at all.

However, all of the expressions before `nil` are still evaluated, so those macro definitions are added to `environment` at compile time.

I'm on the fence about whether this PR is a good idea. It's faster and it allows macros and specials to use local functions, which is nice. But in almost every other Lisp, defining a macro causes that macro become available immediately. If we merge this PR, macros and specials would have to be loaded separately, or wrapped in a `when-compiling` statement, which produces no compiled output. After this PR, I'm not sure there's a nice way to get the current functionality of "define a macro immediately at compile time, but also produce compiled output." (One possibility is to introduce a new macro, `with-eval`, which evaluates its body at compile time like `when-compiling` does, but then returns its body rather than returning the output of the `eval`.)

As an alternative to this PR, it might be better to add functionality like

```
(define-macro foo () :defer
  ...)
```

to indicate that the macro shouldn't be evaluated at compile-time. That would let that particular macro use locally-defined functions, while leaving the default behavior alone. On the other hand, that would add more complexity to the system. So I don't know whether this PR would be best, or whether adding some kind of `:defer` option would be better.

Doing one of those two strategies would probably be good though, since Lumen's current functionality has no way of letting users write macros that use locally-defined functions, and global functions have to already be defined at compile time.